### ISSUE TITLE
Enhancement: Asynchronous database connection.

### ISSUE THREAD

**Issue by AlastairGrowcott** [jtv/libpqxx#487]
Connecting over a slow network link can take some time. Connecting to an unreachable host can also take some time to fail.

My understanding, which may be incorrect, is that constructing a pqxx::connection() blocks while libpqxx attempts to connect to the database. This can cause applications to hang, much to user frustration, and also cause issues in real time data processing applications.

It would be nice if there was a way to construct a pqxx::connection instance without blocking and then access the underlying socket in an event loop to trigger calls to libpqxx to further process the connection, calling a libpqxx function to determine if connecting is successful, in progress, or failed.

I am trying to write an application to track the different versions of firmware that get programmed onto various devices and then send the data back to base (where an internet connection is present) to store it in a database so we know what version of firmware is flashed on each device for each customer. When out in the field we may need to connect using VPN over 3G or 4G - not a fast connection by any means. The last thing I need is my boss getting irate because the software hung for 10 seconds for no apparent reason. Also, detailed error codes as to why connection failed would be good. Was it a network failure or simply wrong credentials? I have looked through the code, but other than trapping exceptions and reporting the exception string there doesn't seem to be any way to determine why a connection failed to open.

**Comment by KayEss** [comment]
The normal solution to the responsiveness problem is to never do anything that might block in a UI thread -- this is good general advice. You should have one (or more) worker threads that perform work and have them pass data back to the UI thread for display.

As for getting the errors, you should have an exception barrier at the right level in your code which can report by looking at the exceptions.

**Comment by AlastairGrowcott** [comment]
So I am a bit new to C++ exceptions - do they report a programmatic code? All I can see is e.what() which returns a user readable string.

Hunting around the documentation, I spotted pqxx::connect_async() in the class reference. This may do what I need.

**Comment by jtv** [comment]
There are no programmatic error codes on connection exceptions at the moment, just the error text.  (You do get programmatic codes on SQL errors, defined by the server and the SQL standard).  It should be possible for libpqxx to obtain more information about a connection failure, it's just that there hasn't been much demand for it.  Usually a connection error means that either the programmer or the user needs to do something differently, and the error message tells them.  There's usually not a lot that you'd want _the application_ to do differently depending on the nature of the failure.

There used to be support for asynchronous connections, and you may have found traces of those.  Here, too, there just wasn't much call for the feature.  I removed those (and a lot of associated complexity) to make room for new improvements.  I'm hoping to do something new that works well with C++20 async.

All that said, I think what @KayEss says makes sense.  Separate threads look like the best option right now.

**Comment by AlastairGrowcott** [comment]
Thanks for the update.

Back in the early noughties, multi-threading was the flavour of the day. People were using it like crazy. My opinion since then has changed little - it has its uses, but they are few and far between and should be avoided where possible.

Still, if, for now, its the only option then I shall go with that.

This thread was only ever intended as a suggestion for an enhancement. If I am the only person asking for it, just put it on the low priority list.

Programmatic error codes are still a good idea though. It allows an application to decide whether to log an error or prompt the user with it.

**Comment by jtv** [comment]
It's a discussion worth having.  I agree that the Java approach of "just use threads for everything" is not a great one.  Limited numbers of workers help, but don't solve everything.  I used to have a separate connection class in libpqxx that started connecting when you created it, but didn't block until you actually started using it.  But it was part of a massively more complicated connection model, and getting rid of that overhead was a real breakthrough.

Nowadays I'm leaning more towards an "async factory" kind of model, with a separate class for an asynchronous connection attempt, which eventually produces a `pqxx::connection` object.  But I'm sort of locked in a chicken-and-egg situation where C++20 async has poor compiler support, and also requires a lot of custom wiring that I need more experience with.  Plus, it poses new questions about division or responsibility — how do we plug libpqxx's model(s) of async waiting (for a connection, for a result) into _your application's_ model of async waiting?

We could probably split categories of connection error into more fine-grained exception classes.  I'll have to dive into the details of where and when these exceptions get thrown.

**Comment by AlastairGrowcott** [comment]
Re: plugging libpqxx's model of async waiting into that of my application, exposing the socket should be enough. It is what I am already using in one of my applications: an event loop where I poll on sockets. On Posix systems it works really well. If you need to do internal processing that doesn't involve socket activity, you can expose one of another type of file descriptor - there are plenty to choose from depending on what you need. It's a bit crap on Windows - I had to write a ton of extra code to implement an interrupt function where I was using a third party library that insisted on using a separate thread, because the interrupt object isn't a socket, but and Windows doesn't really use file descriptors and if you want to wait on a variety of different handles - it is REALLY nasty! And bugged, with the socket handles not working the way the documentation says they should.

Re: Exceptions, I am not keen on having a separate exception class for each failure type. It may be the way C++ does it, but having a string of catch statements where 90% of the code is identical in each is not great. I'd prefer just to switch on the error code as you can group different error codes together where the behaviour is the same, and it just looks neater too.

**Comment by jtv** [comment]
I like the sound of "expose the socket, let the application figure out the waiting."  Of course that does mean that the application needs to know whether it's waiting to send data or not.  (Either way you'd want to wake up when there's incoming data.)

So we could have an API something like...

```cpp
// Ongoing asynchronous connection attempt.
// Use this to produce a pqxx::connection.
// Any exception invalidates the whole object.
struct connecting
{
  connecting(char const connection_string[]);
  ~connecting();
  // Socket.
  int sock() const noexcept;
  // Are we currently waiting to write to the socket?
  bool wait_to_write() const noexcept;
  // Progress towards completion but don't block.
  void communicate();
  // Is our connection finished?
  bool done() const noexcept;
};
```

And once "done" this would match up with a constructor `pqxx::connection::connection(connecting &&)`.  And then hopefully we'd be able to build a coroutine-based wrapper function on top of this later.

**Comment by jtv** [comment]
As for the connection errors, there seems to be very little information we could query, so it'd be hard to come up with a good scheme of error codes.  I think a class hierarchy works better for that kind of situation.

What exact information would you need to know about the failure, that we can provide?  One thing I see we can detect is "failure to provide a password."  But that's a bit of a special case.

**Comment by AlastairGrowcott** [comment]
> 
> 
> I like the sound of "expose the socket, let the application figure out the waiting." Of course that does mean that the application needs to know whether it's waiting to send data or not. (Either way you'd want to wake up when there's incoming data.)
> 
> So we could have an API something like...
> 
> ```c++
> // Ongoing asynchronous connection attempt.
> // Use this to produce a pqxx::connection.
> // Any exception invalidates the whole object.
> struct connecting
> {
>   connecting(char const connection_string[]);
>   ~connecting();
>   // Socket.
>   int sock() const noexcept;
>   // Are we currently waiting to write to the socket?
>   bool wait_to_write() const noexcept;
>   // Progress towards completion but don't block.
>   void communicate();
>   // Is our connection finished?
>   bool done() const noexcept;
> };
> ```
> 
> And once "done" this would match up with a constructor `pqxx::connection::connection(connecting &&)`. And then hopefully we'd be able to build a coroutine-based wrapper function on top of this later.

That looks almost identical to the APIs I use :) I would call it "Process" rather than "communicate" though.

**Comment by AlastairGrowcott** [comment]
> 
> 
> As for the connection errors, there seems to be very little information we could query, so it'd be hard to come up with a good scheme of error codes. I think a class hierarchy works better for that kind of situation.
> 
> What exact information would you need to know about the failure, that we can provide? One thing I see we can detect is "failure to provide a password." But that's a bit of a special case.

Maybe we need a new issue for thie conversation on error codes.

As I see it there are many different error codes but they can roughly be categorised as:

Failed to connect to server (no route to host, host doesn't exist, host refused connection)
Connected to server, but database connection disallowed (server doesn't allow connections - pg_hba.conf)
Failed to authenticate (incorrect username/password)


The reason for the error codes is because the different categories are fixable by different people.

The first is complicated: maybe the user forgot to connect to the internet, maybe they entered in the wrong host name. Alternatively it may be circumstances outside their control, maybe their ISP is having problems or the server has crashed.

The second is a matter for the database administrator - there is nothing the user can do about it (except perhaps try a different username and password if the pg_hba.conf is being that granular).

The third is probably a simple mis-type by the user.

In each case the response of the program should be different. In the first the application can prompt the use to check their network connection and the configured host name and try again. In the second all the application can do is to say that the database is refusing their connection, bye bye, go away. In the third a simple "authentication failed, please try again" response is correct.

At the end of the day it boils down to "can the user fix this problem or do they need to go tell someone else to fix it, and if so who".

**Comment by jtv** [comment]
> As I see it there are many different error codes but they can roughly be categorised as:
> 
> Failed to connect to server (no route to host, host doesn't exist, host refused connection) Connected to server, but database connection disallowed (server doesn't allow connections - pg_hba.conf) Failed to authenticate (incorrect username/password)
> 
> The reason for the error codes is because the different categories are fixable by different people.
> 
> The first is complicated: maybe the user forgot to connect to the internet, maybe they entered in the wrong host name. Alternatively it may be circumstances outside their control, maybe their ISP is having problems or the server has crashed.
> 
> The second is a matter for the database administrator - there is nothing the user can do about it (except perhaps try a different username and password if the pg_hba.conf is being that granular).
> 
> The third is probably a simple mis-type by the user.
> 
> In each case the response of the program should be different. In the first the application can prompt the use to check their network connection and the configured host name and try again. In the second all the application can do is to say that the database is refusing their connection, bye bye, go away. In the third a simple "authentication failed, please try again" response is correct.
> 
> At the end of the day it boils down to "can the user fix this problem or do they need to go tell someone else to fix it, and if so who".

It seems to me like we've got several factors at play here:
 * What's the root cause of the error?
 * Who should address the problem?
 * What can the application do to help?

The "Who" question is the least important to me, because regardless of who's responsible for fixing the problem, the application's main job is still to tell the user and the operators about it.  Operators can keep a statistical watch on errors to look for sudden spikes that may indicate bugs, server problems, or other infrastructural failures.  But this area is so diverse that you've really got to deal with it as text pattern recognition.

Also, my experience is that the categories you get from querying error properties tend to correlate very poorly to root causes or things the application can do.  I guess the infrastructure will usually handle the really clear-cut cases already.

There are exceptions of course.  Sometimes you know exactly what the application can do about some very specific problem.  So I'm especially interested in those examples: "Here's something I would make my application do if only I knew X about a given error."  It's a very opportunistic game, and as I think you were saying, there probably isn't much to be gained from clever taxonomies.

**Comment by AlastairGrowcott** [comment]
Actually I was saying that there is a lot to be gained.

In the application I am currently working on, the connection is kind of hidden from the user. I am uploading, effectively, usage data to the database. The system the application is running on might very well not have an internet connection at all, in which case the data will be written to file for later upload. So...

1. If there is an outright failure to connect: the application flags this silently and writes data to file.
2. If connection is successful but refused by the database (pg_hba.conf): the application needs to get an error report to the database administrator (automated email).
3. If the connection is successful but authentication fails: the application needs to get an error report to the application developer to update the database credentials in the application. Alternatively the application may be able to check if there is a newer version available and advise the user to do an upgrade.

Parsing textual error messages is a pain, not least because the messages may change from one version of PostgreSQL/libpq/libpqxx to another. Error codes generally don't because programmers are aware of the pain that can cause. They may obsolete error codes and add new ones, but by and large they remain the same.

**Comment by AlastairGrowcott** [comment]
As a follow on to my previous post, one of the things I hate most in life is when people, including most especially my 8 year old son, say "it doesn't work!" The response to this is "what doesn't work?" and "In what way doesn't it work?". The text messages contain this information, but it is useful to the application to know too and not simply receive a generic exception without that information.

**Comment by AlastairGrowcott** [comment]
We really need to move the conversation about error codes to another thread or another medium. I am happy to exchange emails directly if you like.

**Comment by jtv** [comment]
Totally agree that we need to make the nature of the problem clear to the hapless humans.  Sure, I'd say just file a new issue for this.

Meanwhile I'll see about getting some of the async connection work done.  I'm not well at the moment, so won't be world quality at first.

**Comment by AlastairGrowcott** [comment]
There's no hurry. It was a suggestion for an improvement, not a demand!

Wait until you are well. I hope you get better soon.

**Comment by jtv** [comment]
What can I say — lying down all day can be a bit boring.  So I just pushed an implementation.  See `pqxx::connecting`, in `pqxx/connection.hxx`.

**Comment by jtv** [comment]
@AlastairGrowcott any chance you could give the new `pqxx::connecting` a whirl?

I can't decide whether the naming is incredibly dumb or total genius, so if you have a much better idea... last chance.  :-)

**Comment by AlastairGrowcott** [comment]
Hi.

I'll do what I can but as of the end of last week I got moved from what I was doing to something else - where I need to try and do two weeks work in one week.

**Comment by AlastairGrowcott** [comment]
I have trouble using libpqxx with persistent member variables. My C++ is not the greatest. Normally I use pointers, but since produce() returns an instance not a pointer, and there does not appear to be a copy constructor, I am a bit stuck. I suppose I could have a pointer to the "pqxx::connecting" pointer and then call produce() every time I need to access the actual connection, but that seems a bit...poor.

How do you recommend I store a connection as a member variable? Like this?

    my_class.h
    ---------------
    ...
    pqxx::connecting m_connecting;
    pqxx::connection m_connection;
    bool                     m_connected;

    my_class.cpp
    ------------------
    MyClass::MyClass() : m_connecting(), m_connection(), m_connected{false}
    {
    }

    MyClass::Start()
    {
        m_connecting = pqxx::connecting("my.db.local");
    }

    MyClass::Process()
    {
        if (!m_connected) {
            m_connecting.process();
            if (m_connecting.done()) {
                **m_connection = std::move(m_connecting.produce());**
                m_connected = true;
            }
        } else {
            ...
        }
    }

**Comment by jtv** [comment]
The line you highlighted looks right to me, yes.  You can't copy a connection, but you can move it.  You can only call `produce()` once; it is intended that you store the connection object at that point.

(Of course in real life you'll need to call `m_connecting.process()` etc. in a loop until done, but I guess this is just example code.)

**Comment by jtv** [comment]
The normal synchronous connection constructors _do_ throw exceptions.  To be precise, they call `init()` which calls `complete_init()` which checks whether the connection succeeded.  If it didn't, `complete_init()` throws `broken_connection`.

The idea of putting the asynchronous connection process inside the connection is exactly what I got rid of in the first place.  It complicates the internal design and introduces additional states which may affect your application.

**Comment by AlastairGrowcott** [comment]
MyClass::Process will be called by the event loop which will poll the libpqxx socket along with all other sockets, and when the poll returns call Process() functions on everything in the system.

Which means I also need to be careful that I get the socket from the right object, as well as the flags to check for read and check for write.

This is feeling more and more complicated all the time.

**Comment by jtv** [comment]
Yes, that's all in the nature of async programming I'm afraid.  The fundamental problem is that you need to turn your abstractions "inside out" so that your application can take over the parts where you wait for things.  And when you wait for multiple things, there's no getting around having to make sure that you're waiting for the right things — even though those details may rightfully belong buried deep down in lower abstractions.

I wish I could at least guarantee that the socket wouldn't change mid-connect.  But libpq does not guarantee that.  I could perhaps throw an error when `connecting` detects a socket change, but even then you'd still need to get the "wait for read/wait for write" bits on every iteration.  So it probably wouldn't help much.

**Comment by AlastairGrowcott** [comment]
> Also, what I think about pqxx async implementation. It should be modern C++. Modern == templated. Templated in such way that user will control underlying IO whatever he wants. Also I see that many functions has small bodies, so they can be moved into class itself, making it easier to make pqxx even header only. Or introduce more templated code at least.

Rule number 1 of good code: Never do anything for the sake of doing it. Why should libpq++ be more templated? Just to be more modern? That is a bad reason. Because there is a specific design rationale to do so? That is a good reason.

Templates are not necessarily better. Header only libraries are not necessarily better (in many cases they are much much worse).

I have been a software engineer for 25 years and so many times I have seen people assuming that newer=better. It isn't necessarily so. C is one of the oldest programming languages and it is still very very much in use as the best choice for certain applications. Someone on Stack Overflow once commented about C being a dead/obsolete language. It is very much not dead or obsolete.

**Comment by AlastairGrowcott** [comment]
This may not be the place for this rant, but, I just remembered...

I am using a WebSocket C++ library. It is VERY "modern", very heavily templated, header only. It has a resource leak. Because it is so heavily templated and header only, it is inordinately complicated which makes it extremely hard to maintain or debug. It also uses asio. I have made a couple of attempts to look at the resource leak, as well as look into other issues I have had with this library, and been unable to make much headway.

Simpler = better. You write code once, but it gets read (and modified) many times. The simpler the code is, the quicker it is to understand and the quicker changes will be to make, and with a better quality to the changes so that you are less likely to introduce bugs. More templating means more complexity. Header only libraries mean more complexity.

**Comment by jtv** [comment]
Totally agree: software often walks a razor's edge between features and simplicity.

For now though I don't think there's anything we need to template.  Even better than templating how you wait for a socket is to leave it outside of libpqxx.  However I do see one thing I can improve there: the documentation for `await_notification` should describe how you can do your own waiting.

One of the reasons why libpqxx is not a header-only library, by the way, is that it encapsulates libpq.  To compile your libpqxx-based application you do not need the libpq headers.  You only need the library binary.  It's a small thing, but I like it.

So I'm going the other way and betting on C++ _modules._  Really looking forward to those.  Hopefully they'll speed up compilation.  But it'll be a major rework once we're comfortable with the available compiler support.

**Comment by jtv** [comment]
Anyway, I think we can close this ticket now.  If there are any problems with the new class, that'll be for a new ticket.

**Comment by jtv** [comment]
@AlastairGrowcott @egorpugin @KayEss I'm making an incompatible change to the `connecting` class, since I haven't formally released it yet!

The member function `pqxx::connecting::produce()` will become rvalue-qualified.  This means that you will only be able to invoke it on an rvalue, expressing the fact that the `connecting` object becomes useless at that point.

**Comment by AlastairGrowcott** [comment]
Sounds good. I still haven't found time to experiment with the new class. Sorry.

**Comment by KayEss** [comment]
I've updated the example project. You can see how the change looks in situ here: https://github.com/KayEss/pqxx-async/blob/main/examples/simple.cpp#L23

**Comment by jtv** [comment]
Excellent.  That project also creates a bridge to C++ coroutines.

If the API for this turns out to be a keeper, I'd like to build similar
provisions for async querying.

**Comment by KayEss** [comment]
Indeed it does :-) I'm just starting a branch to look at io_uring support
